diff --git a/src/com/android/nfc/NfcDispatcher.java b/src/com/android/nfc/NfcDispatcher.java
index 1721d1a..3eda1ec 100644
--- a/src/com/android/nfc/NfcDispatcher.java
+++ b/src/com/android/nfc/NfcDispatcher.java
@@ -16,9 +16,6 @@
 
 package com.android.nfc;
 
-import com.android.nfc.RegisteredComponentCache.ComponentInfo;
-import com.android.nfc.handover.HandoverManager;
-
 import android.app.Activity;
 import android.app.ActivityManager;
 import android.app.ActivityManagerNative;
@@ -34,15 +31,22 @@ import android.content.pm.PackageManager;
 import android.content.pm.PackageManager.NameNotFoundException;
 import android.content.pm.ResolveInfo;
 import android.net.Uri;
+import android.nfc.IPhoneKey;
 import android.nfc.NdefMessage;
 import android.nfc.NdefRecord;
 import android.nfc.NfcAdapter;
 import android.nfc.Tag;
+import android.nfc.tech.MifareUltralight;
 import android.nfc.tech.Ndef;
+import android.os.IBinder;
 import android.os.RemoteException;
+import android.os.ServiceManager;
 import android.os.UserHandle;
 import android.util.Log;
 
+import com.android.nfc.RegisteredComponentCache.ComponentInfo;
+import com.android.nfc.handover.HandoverManager;
+
 import java.io.FileDescriptor;
 import java.io.PrintWriter;
 import java.nio.charset.Charsets;
@@ -68,6 +72,8 @@ public class NfcDispatcher {
     PendingIntent mOverrideIntent;
     IntentFilter[] mOverrideFilters;
     String[][] mOverrideTechLists;
+    
+    IPhoneKey phoneKeyService;
 
     public NfcDispatcher(Context context, HandoverManager handoverManager) {
         mContext = context;
@@ -76,6 +82,7 @@ public class NfcDispatcher {
                 NfcAdapter.ACTION_TECH_DISCOVERED, NfcAdapter.ACTION_TECH_DISCOVERED);
         mContentResolver = context.getContentResolver();
         mHandoverManager = handoverManager;
+        phoneKeyService = getServiceInterface();
     }
 
     public synchronized void setForegroundDispatch(PendingIntent intent,
@@ -198,6 +205,18 @@ public class NfcDispatcher {
             overrideIntent = mOverrideIntent;
             overrideTechLists = mOverrideTechLists;
         }
+        
+        
+        try {
+            boolean isMath = isTechMatch(tag, new String[][] { new String[] { MifareUltralight.class.getName()} });
+            if(phoneKeyService !=null && isMath && phoneKeyService.isEnable()){
+                phoneKeyService.handleTag(tag);
+                return true;
+            }
+        } catch (RemoteException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
 
         resumeAppSwitches();
 
@@ -494,4 +513,12 @@ public class NfcDispatcher {
             pw.println("mOverrideTechLists=" + mOverrideTechLists);
         }
     }
+    
+    private IPhoneKey getServiceInterface(){
+        IBinder binder = ServiceManager.getService("phonekey");
+        if(binder == null)
+            return null;
+        
+        return IPhoneKey.Stub.asInterface(binder);
+    }
 }
